Model viewUrl query parameters

Allcharles opened the pull request:
# Update viewUrl to handle query paramaters

Updated how the `viewUrl` for models works, and how they interact with the angular router.

## Changes

- Created `UriDirective`
- Updated models to return viewUrls including query parameters

## Issues

Closes #772 

## Final Checklist

- [x] Assign reviewers if you have permission
- [x] Assign labels if you have permission
- [x] Link issues related to PR
- [x] Ensure project linter is not producing any warnings (`npm run lint`)
- [x] Ensure build is passing on all browsers (`npm run test:all`)
- [x] Ensure CI build is passing
- [x] Ensure docker container is passing ([docs](https://github.com/QutEcoacoustics/workbench-client#docker))

Allcharles committed:
Update model viewUrls

Allcharles committed:
Update admin pages

Allcharles committed:
Created uri directive

Allcharles committed:
Functioning uri directive

Allcharles committed:
Replace routerLink with uri

Allcharles committed:
Cleanup compile errors

Allcharles committed:
Simplify item component unit tests

Allcharles committed:
Updated tests

Allcharles committed:
Created uri directive tests

atruskie commented on src/app/test/helpers/html.ts:
```diff
@@ -99,7 +98,20 @@ export function assertHref(
+export function assertUri(target: HTMLElement, route: string) {
+  expect(target).toBeTruthy("No route detected");
+  expect(target).toHaveAttribute("ng-reflect-uri");
+  expect(target).toHaveAttribute("href", encodeURI(route));
```
rename uri to url

atruskie commented on src/app/models/Tag.ts:
```diff
@@ -59,10 +59,7 @@ export class Tag extends AbstractModel implements ITag {
-    // /library?reference=all&tagsPartial=[tag.text]
-    // TODO This link is not working (issue #772)
-    return libraryMenuItem.route.toRouterLink({
+    return libraryMenuItem.route.format(undefined, {
```
this `undefined` seems strange - is there a reason for it?

I'm going to guess we're omitting a route parameter? But which one?

Allcharles replied:
`libraryMenuItem` has no route parameters

atruskie commented on src/app/models/Script.ts:
```diff
@@ -71,6 +71,11 @@ export class Script extends AbstractModel implements IScript {
   public get viewUrl(): string {
-    return adminScriptMenuItem.route.toRouterLink({ scriptId: this.id });
+    console.warn("Script viewUrl not implemented.");
+    return "not_implemented";
```
throw - this will just lead to a _misleading_ 404

Is there a reason we can't just return the properly formed url? It will still 404 but at least it is self fixing when the associated view is eventually created

atruskie commented on src/app/models/SavedSearch.ts:
```diff
@@ -50,7 +50,7 @@ export class SavedSearch extends AbstractModel implements ISavedSearch {
   public get viewUrl(): string {
-    console.error("SavedSearch viewUrl not implement");
+    console.warn("SavedSearch viewUrl not implemented.");
     return "not_implemented";
```
See comments on script.ts - should not return "not_implemented"

atruskie commented on src/app/models/AbstractModel.ts:
```diff
@@ -42,6 +43,7 @@ export abstract class AbstractModel {
   public readonly kind: string;
 
   /**
+   * TODO Update comment
```
carpe diem

atruskie commented on src/app/directives/uri/uri.directive.ts:
```diff
@@ -0,0 +1,51 @@
+
+@Directive({
+  // eslint-disable-next-line @angular-eslint/directive-selector
+  selector: "a[uri]",
```
`uri` ➡`url`

atruskie replied:
I know we talked about documenting over namespacing, but https://angular.io/guide/styleguide#directive-custom-prefix is a compelling argument. 

Suggested `bawUrl`

atruskie commented on src/app/directives/uri/uri.directive.ts:
```diff
@@ -0,0 +1,51 @@
+  selector: "a[uri]",
+})
+export class UriDirective extends withUnsubscribe(RouterLinkWithHref) {
+  @Input() public uri: string;
```
Does not match specification from #772: `string | UrlTree`

atruskie commented on src/app/directives/uri/uri.directive.ts:
```diff
@@ -0,0 +1,51 @@
+    const queryParams: Params = {};
+    url.searchParams.forEach((value, key) => (queryParams[key] = value));
+
+    return this._router.createUrlTree([url.pathname], {
```
Why not: [`parseUrl`](https://angular.io/api/router/Router#parseurl) as defined in spec?

It handles the all the cases you handle manually: https://github.com/angular/angular/blob/e3f09ce845cef9fd2c89d39d0d822114c51e68fa/packages/router/src/url_tree.ts#L289-L294

atruskie replied:
You can use object.assign to then merge in the base directive params with the created urlTree

atruskie commented on src/app/directives/strongRoute/strong-route.directive.ts:
```diff
@@ -67,7 +63,11 @@ export class StrongRouteDirective
     });
   }
 }
+
+function attrBoolValue(s: any): boolean {
+  return s === "" || !!s;
+}
```
repeated function in uri.directive.ts

atruskie reviewed (changes requested):
Fantastic work. 

Minor changes:
-  `uri` should be `url`
- We should add a section to contributing.md that explains what custom directives are in use (particularly since these could appear magical or external/angular supplied directives)
- You used the `string` and `format` pairing a lot. Is there a reason for that over `UrlTree` and `formatUrlTree`? Technically it's more work for the browser - though it is unlikely to be a significant performance penalty
- I think you need a consistent approach for dealing with urls that are not yet ready. In this review I saw:
    - throwing an exception
    - returning `""`
    - returning `"not_implemented"`
    - returning `""/play_url"`
    - various combinations of the above with and without console errors or warns

Allcharles committed:
Rename uri to url

Allcharles committed:
Modified UrlDirective prefix to use bawUrl

Allcharles committed:
Update AbstractModel comments

Allcharles committed:
Simplify urlTree method

Allcharles committed:
Make model viewUrls error out

Updated render-view to handle viewUrl's throwing errors so that pages/tests wont crash

Allcharles committed:
Updated render-field component tests

Allcharles committed:
Updated comment to highlight potential danger

Allcharles committed:
Merge branch 'master' into viewUrl-cleanup

atruskie reviewed (approved):
Perfect
